### PR TITLE
Expose "parsingAndImporting" setting to user

### LINKED ISSUE #11756: `absolutePath` should be available for imports in the AST when using `stopAfter: parsing`
## Abstract
Tools should be using `absolutePath` from `ImportDirective` AST node for more reliable import resolution but it's not available when `stopAfter: parsing` option is used - even though that's when they need it the most.

## Motivation
One of the uses case for `stopAfter: parsing` is to make it possible for tools to easily scan source files for imports. The raw path used in the import statement however is of limited utility - many tools are actually interested in a resolved path that can be used as a key in `sources` in Standard JSON. This resolution must account for [relative imports](https://docs.soliditylang.org/en/latest/path-resolution.html#relative-imports) and [remappings](https://docs.soliditylang.org/en/latest/path-resolution.html#import-remapping) and current compiler logic for doing that is full of unintuitive corner cases so it's easy to get it wrong. Doing it wrong, however, results in the compiler not being able to find the file matching the import where it expects it to be.

The compiler does make the resolved path available in the AST but it's not there yet when `stopAfter: parsing` setting is used.

## Specification
Either perform import resolution earlier so that `absolutePath` is already in the AST when `stopAfter: parsing` is used or introduce a new value for `stopAfter` (`parsingAndImportResolution`?) that still stops before analysis but after import resolution.

## Backwards Compatibility
Making a previously unavailable AST key available is fully backwards-compatible.

#### Input
```json
{
    "language": "Solidity",
    "sources": {
        "/project/../C.sol": {"content": "import \"../L.sol\";"}
    },
    "settings": {
        "stopAfter": "parsing",
        "remappings": [":/project/=/lib/"],
        "outputSelection": {"*": {"": ["ast"]}}
    }
}
```
#### AST from the output
```json
{
    "sources": {
        "/project/../C.sol": {
            "ast": {
                "absolutePath": "/project/../C.sol",
                "id": 2,
                "nodeType": "SourceUnit",
                "nodes": [
                    {
                        "file": "../L.sol",
                        "id": 1,
                        "nameLocation": "-1:-1:-1",
                        "nodeType": "ImportDirective",
                        "src": "0:18:0",
                        "symbolAliases": [],
                        "unitAlias": ""
                    }
                ],
                "src": "0:18:0"
            },
            "id": 0
        }
    }
}
```

#### Input
```json
{
    "language": "Solidity",
    "sources": {
        "/project/../C.sol": {"content": "import \"../L.sol\";"},
        "/lib/L.sol": {"content": ""}
    },
    "settings": {
        "remappings": [":/project/=/lib/"],
        "outputSelection": {"*": {"": ["ast"]}}
    }
}
```
#### AST from the output
```json
{
    "sources": {
        "/lib/L.sol": {
            "ast": {
                "absolutePath": "/lib/L.sol",
                "exportedSymbols": {},
                "id": 1,
                "nodeType": "SourceUnit",
                "nodes": [],
                "src": "0:0:0"
            },
            "id": 0
        },
        "/project/../C.sol": {
            "ast": {
                "absolutePath": "/project/../C.sol",
                "exportedSymbols": {},
                "id": 3,
                "nodeType": "SourceUnit",
                "nodes": [
                    {
                        "absolutePath": "/lib/L.sol",
                        "file": "../L.sol",
                        "id": 2,
                        "nameLocation": "-1:-1:-1",
                        "nodeType": "ImportDirective",
                        "scope": 3,
                        "sourceUnit": 1,
                        "src": "0:18:1",
                        "symbolAliases": [],
                        "unitAlias": ""
                    }
                ],
                "src": "0:18:1"
            },
            "id": 1
        }
    }
}
```

### PR DESCRIPTION
**Pull request by Marenz**
fixes #11756

### CODE REVIEW

**Comment by chriseth** [comment]
I don't understand - why do we need a new "stop after" setting? I thought the idea would be to just fill the absolute path during parsing and before actually trying to load the file?

**Comment by Marenz** [comment]
The ticket specified 

> Either perform import resolution earlier so that absolutePath is already in the AST when stopAfter: parsing is used or introduce a new value for stopAfter (parsingAndImportResolution?) that still stops before analysis but after import resolution.

and I figured we have that internal step already so just exposing it would be the easiest solution.. and I figured it makes sense to expose more steps as necessary..

**Comment by chriseth** [comment]
I think that, unless it is complicated to do, we should expose more useful data earlier.

File: libsolidity/interface/CompilerStack.cpp [hunk 1]
```diff
@@ -1102,12 +1116,8 @@ StringMap CompilerStack::loadMissingSources(SourceUnit const& _ast, std::string
 			{
 				solAssert(!import->path().empty(), "Import path cannot be empty.");
```
Comment (chriseth) [inline libsolidity/interface/CompilerStack.cpp]: Can you check for `import->annotation().absolutePath` instead now?

File: libsolidity/interface/CompilerStack.cpp [hunk 2]
```diff
@@ -350,8 +350,22 @@ bool CompilerStack::parse()
 		else
 		{
 			source.ast->annotation().path = path;
+
+			for (auto const& node: source.ast->nodes())
```
Comment (chriseth) [inline libsolidity/interface/CompilerStack.cpp]: You can use `filteredNodes` here

File: libsolidity/interface/CompilerStack.cpp [hunk 3]
```diff
@@ -350,8 +350,22 @@ bool CompilerStack::parse()
...
+				if (ImportDirective const* import = dynamic_cast<ImportDirective*>(node.get()))
+				{
+					solAssert(!import->path().empty(), "Import path cannot be empty.");
+
+					string importPath = util::absolutePath(import->path(), path);
+					// The current value of `path` is the absolute path as seen from this source file.
+					// We first have to apply remappings before we can store the actual absolute path
+					// as seen globally.
+					importPath = applyRemapping(importPath, path);
+					import->annotation().absolutePath = importPath;
```
Comment (chriseth) [inline libsolidity/interface/CompilerStack.cpp]:
```suggestion
					import->annotation().absolutePath = applyRemapping(importPath, path);
```

File: Changelog.md [hunk 4]
```diff
@@ -8,6 +8,7 @@ Compiler Features:
  * Commandline Interface: Accept nested brackets in step sequences passed to ``--yul-optimizations``.
  * Commandline Interface: Add ``--debug-info`` option for selecting how much extra debug information should be included in the produced EVM assembly and Yul code.
  * Commandline Interface: Use different colors when printing errors, warnings and infos.
+ * General: Absolute paths of imports are now set in the ``parsing`` stage.
```
Comment (chriseth) [inline Changelog.md]: We could use "now" in every entry, so I think it is better to not use it at all.

Comment (chriseth) [reply Changelog.md]: Oh and as category, maybe "JSON AST"? Because that is what is relevant to the user reading the changelog.

Comment (Marenz) [reply Changelog.md]: The option is available using the command line as well though?

Comment (chriseth) [reply Changelog.md]: Sure, but the json ast is a generic thing that is available to all ways to access the compiler.

**Comment by leonardoalt** [comment]
Tests failing

File: libsolidity/interface/CompilerStack.cpp (hunk 1 above)
Comment (Marenz) [reply libsolidity/interface/CompilerStack.cpp]: It turns out: No. The path can actually be empty after we resolved it. What was previously "." becomes empty. But we still have the assert pre-resolve stage as it was originally...

File: libsolidity/interface/CompilerStack.cpp [hunk 5]
```diff
@@ -350,8 +350,26 @@ bool CompilerStack::parse()
...
+					// as seen globally.
+					import->annotation().absolutePath =
+						applyRemapping(
+							util::absolutePath(
+								import->path(),
+								path
+							),
+							path
+						);
+				}
```
Comment (chriseth) [inline libsolidity/interface/CompilerStack.cpp]:
```suggestion
			for (auto const& import: ASTNode::filteredNodes<ImportDirective>(source.ast->nodes()))
			{
				solAssert(!import->path().empty(), "Import path cannot be empty.");

				// The current value of `path` is the absolute path as seen from this source file.
				// We first have to apply remappings before we can store the actual absolute path
				// as seen globally.
				import->annotation().absolutePath = applyRemapping(util::absolutePath(
					import->path(),
					path
				), path);
			}
```

File: test/cmdlineTests/standard_parsing_import_absolute_paths/output.json [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+                "nodes": [],
+                "src": "0:0:0"
+            },
+            "id": 0
+        },
+        "/project/../C.sol":
+        {
+            "ast":
+            {
+                "absolutePath": "/project/../C.sol",
```
Comment (chriseth) [inline test/cmdlineTests/standard_parsing_import_absolute_paths/output.json:34]: Interesting!

Comment (chriseth) [reply test/cmdlineTests/standard_parsing_import_absolute_paths/output.json:34]: Ah ok, it was named like that in the input json.

**Review by chriseth — DISMISSED** [review]

**Review by chriseth — APPROVED** [review]

**Pull request merged**